Fix/simplify exception handling

martin-cs opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

We had a half implemented good idea in parse_options_baset in which the main method would catch all of the various exceptions and do a standard "print and then exit with the right code" handling so that it didn't need to be duplicated in each binary.  The first patch extends this so it really does catch all exceptions, the rest of the patches (ordered in increasing intrusiveness / objectionableness) then remove the redundant (and in many cases inconsistent) exception handling from various main program loops.  One thing that isn't done is removing the block parts of the try as it results in large, kinda pointless whitespace commits.  I'll do them if people want.

Commit:
Add default handling for the kinds of exceptions used in CPROVER

This makes the code reflect the comment "catch all exceptions
here so that this code is not duplicated for each tool".

Commit:
Remove redundant try/catch blocks from goto-analyzer

Using the previous commit these should increase the consistency
of how exceptions are handled.

Commit:
Remove redundant try/catch blocks from cbmc

Again, these should only result in a net improvement in how
exceptions are handled; return codes shouldn't change.

Commit:
Remove redundant try/catch blocks from goto-instrument

This does change the exit codes of goto-instrument in the case of
exceptions being thrown.  Doing so eliminates a long-standing
anomoly where goto-instrument had a different exit code for
exceptions to all of the other tools.

Commit:
Remove some exception handling in goto-analyzer

This changes the return value in a few exceptional cases in favour
of being clearer about when an exception is thrown and simplifying
the code.

Commit:
Remove some exception handling in CBMC

This changes the return value in a few exceptional cases but is
more consistent and simpler.

Commit:
Remove some exception handling in goto-diff

Removes specific handling of a few kinds of exception in favour
of the general handlers in parse_options_baset.  This will change
the exit code in a few exceptional cases but makes the code
simpler and more consistent.

tautschnig commented on src/util/parse_options.cpp:
```diff
@@ -71,16 +71,54 @@ int parse_options_baset::main()
+  {
+    std::cerr << e.what() << '\n';
+    return CPROVER_EXIT_EXCEPTION;
+  }
```
Is this case actually needed? Shouldn't the `cprover_exception_baset` below catch all of them?

martin-cs replied:
I don't think so but it was an explicit case in the code I removed so I thought that it had been separated out for a reason and that reason should be preserved.

tautschnig replied:
Fair enough, but then you might add an extra commit that removes it? You've got a catch-all at the very end anyway.

martin-cs replied:
I've added a commit that does this.  Please merge if you are happy with this.

hannes-steffenhagen-diffblue replied:
@martin-cs Historical note: When it was created there originally wasn't a base class, so we had separate handlers for each exception. A base class was added by vote. Also, we weren't entirely sure if all kinds of exceptions used the same exit code. Both of these situations were eventually remedied, so I don't think it's necessary anymore. This probably slipped past us when the situation changed.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5804cb0f3cded517388b889554932855d4fffad0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98315938

Commit:
Remove redundant case in the exception handling in parse_options.cpp

For ease of migration this case was kept from the calling code.
However it is actually redundant and so is removed in this commit.

hannes-steffenhagen-diffblue commented on src/util/exit_codes.h:
```diff
@@ -40,7 +40,6 @@ Author: Martin Brain, martin.brain@diffblue.com
 /// An (unanticipated) exception was thrown during computation.
 #define CPROVER_EXIT_EXCEPTION 6
 // should contemplate EX_SOFTWARE from sysexits.h
-#define CPROVER_EXIT_EXCEPTION_GOTO_INSTRUMENT 11
```
This will lead to a change in exit code for `goto-instrument`. That doesn't bother me personally, but it's a change in behaviour, not sure if it needs to documented somewhere. Also, the comment above should be removed together with this `#define`

martin-cs replied:
1. It is documented in the commit message.

2. The comment is general to all `CPROVER_EXIT_EXCEPTION`, not specifically this one.

hannes-steffenhagen-diffblue commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -555,8 +555,8 @@ int goto_analyzer_parse_optionst::perform_analysis(const optionst &options)
-  if(set_properties())
-    return CPROVER_EXIT_SET_PROPERTIES_FAILED;
+  if(cmdline.isset("property"))
+    ::set_properties(goto_model, cmdline.get_values("property"));
```
What does this change mean?

martin-cs replied:
After removing the `try` block, `set_properties` only contained the two lines I've introduced and then `return false` so I inlined them.

hannes-steffenhagen-diffblue reviewed (approved):
LGTM modulo minor comments

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2212653467140354f2548ff4a46eaef7cf06873a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98344221

kroening commented:
Will this work at all with --json-ui, or --xml-ui? No error reporting for those anymore?

martin-cs commented:
I think the exception text will be printed to std::err rather than message() but TBH it was a bit of a mish-mash where it would go before this patch.

kroening commented:
That needs to be sorted -- all UI integrations I know rely on these showing up in the JSON.

kroening commented:
(We should have had a test...)

martin-cs commented:
@kroening : I'll work on a patch to make sure the we always get the exception text to the right output channel.

It seems there are two ways of doing this : have `parse_optionst` inherit from `messaget` or have a call-back in `parse_optionst` for outputting messages.  Do you have any preference?

martin-cs commented:
PR to allow redirection is here : https://github.com/diffblue/gnat2goto/pull/113
